spring-integration-http/src/main/java/org/springframework/integration/http/inbound/ReactiveHttpInboundEndpoint.java

```diff
@@ -0,0 +1,382 @@
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import org.springframework.http.server.reactive.ServerHttpResponse;
+import org.springframework.integration.gateway.MessagingGatewaySupport;
+import org.springframework.integration.support.AbstractIntegrationMessageBuilder;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.reactive.HandlerMapping;
+import org.springframework.web.reactive.accept.HeaderContentTypeResolver;
+import org.springframework.web.reactive.accept.RequestedContentTypeResolver;
+import org.springframework.web.server.NotAcceptableStatusException;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.WebHandler;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * A {@link MessagingGatewaySupport} implementation for Spring WebFlux
+ * HTTP requests execution.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.0
+ *
+ * @see org.springframework.web.reactive.result.HandlerResultHandlerSupport
+ */
+public class ReactiveHttpInboundEndpoint extends BaseHttpInboundEndpoint implements WebHandler {
+
+	private static final MediaType MEDIA_TYPE_APPLICATION_ALL = new MediaType("application");
+
+	private ServerCodecConfigurer codecConfigurer = ServerCodecConfigurer.create();
+
+	private RequestedContentTypeResolver requestedContentTypeResolver = new HeaderContentTypeResolver();
+
+	public ReactiveHttpInboundEndpoint() {
+		this(true);
+	}
+
+	public ReactiveHttpInboundEndpoint(boolean expectReply) {
+		super(expectReply);
+	}
+
+	public void setCodecConfigurer(ServerCodecConfigurer codecConfigurer) {
+		Assert.notNull(codecConfigurer, "'codecConfigurer' must not be null");
+		this.codecConfigurer = codecConfigurer;
+	}
+
+	public void setRequestedContentTypeResolver(RequestedContentTypeResolver requestedContentTypeResolver) {
+		Assert.notNull(requestedContentTypeResolver, "'requestedContentTypeResolver' must not be null");
+		this.requestedContentTypeResolver = requestedContentTypeResolver;
+	}
+
+	@Override
+	public String getComponentType() {
+		return super.getComponentType().replaceFirst("(http:)", "$1reactive-");
+	}
+
+	@Override
+	protected void onInit() throws Exception {
+		super.onInit();
+	}
+
+	@Override
+	public Mono<Void> handle(ServerWebExchange exchange) {
+		return Mono.defer(() -> {
+			if (isRunning()) {
+				return doHandle(exchange);
+			}
+			else {
+				return serviceUnavailableResponse(exchange);
+			}
+		});
+	}
+
+	@SuppressWarnings("unchecked")
+	private Mono<Void> doHandle(ServerWebExchange exchange) {
+		return Mono.just(exchange)
+				.doOnSubscribe(s -> this.activeCount.incrementAndGet())
+				.flatMap(this::extractRequestBody)
+				.switchIfEmpty(Mono.just(exchange.getRequest().getQueryParams()))
```

Comment:
Also the `switchIfEmpty` is probably better handled as part of `extractRequestBody` where there are some checks already for what kind of request it is. For example an HTTP POST with form data should always have a body and hence a fallback on query params doesn't make much sense. And do you really want query params as the body any time there is no body? I would think that only makes sense with something like a search form, i.e. HTTP GET with query params.